### PR TITLE
add support for property aggregation

### PR DESCRIPTION
**Pull request by akphi**
#### What type of PR is this?

New Feature

#### What does this PR do / why is it needed ?

Added support for property aggregation: `AggregationKind.None`, `AggregationKind.Shared`, `AggregationKind.Composite` like what we have in `legend-pure`

### CODE REVIEW

File: legend-engine-language-pure-grammar/src/test/java/org/finos/legend/engine/language/pure/grammar/test/roundtrip/TestDomainGrammarRoundtrip.java [hunk 1]
```diff
@@ -19,6 +19,28 @@
 
 public class TestDomainGrammarRoundtrip extends TestGrammarRoundtrip.TestGrammarRoundtripTestSuite
 {
+    @Test
```
Comment (epsstan) [inline legend-engine-language-pure-grammar/src/test/java/org/finos/legend/engine/language/pure/grammar/test/roundtrip/TestDomainGrammarRoundtrip.java:22]: Not sure about this grammar.  Any reason why we this should not be modeled as a profile/tag value ? 

@pierredebelen might have a strong view on this one.

Generally, my preference is to keep the variations in syntax/grammar to a minimum.

Comment (akphi) [reply legend-engine-language-pure-grammar/src/test/java/org/finos/legend/engine/language/pure/grammar/test/roundtrip/TestDomainGrammarRoundtrip.java:22]: I bring over what we support in Pure, don't think we should diverge here. Also this is a native UML capability

File: legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/HelperModelBuilder.java [hunk 2]
```diff
@@ -77,6 +79,14 @@ public static org.eclipse.collections.api.block.function.Function<Property, org.
...
                     ._stereotypes(ListIterate.collect(property.stereotypes, s -> context.resolveStereotype(s.profile, s.value, s.profileSourceInformation, s.sourceInformation)))
                     ._taggedValues(ListIterate.collect(property.taggedValues, t -> new Root_meta_pure_metamodel_extension_TaggedValue_Impl("", null, context.pureModel.getClass("meta::pure::metamodel::extension::TaggedValue"))._tag(context.resolveTag(t.tag.profile, t.tag.value, t.tag.profileSourceInformation, t.sourceInformation))._value(t.value)))
+                    ._aggregation(AggregationKind.COMPOSITE.equals(property.aggregation)
+                            ? context.pureModel.getEnumValue(M3Paths.AggregationKind, "Composite")
+                            : AggregationKind.SHARED.equals(property.aggregation)
+                            ? context.pureModel.getEnumValue(M3Paths.AggregationKind, "Shared")
+                            : AggregationKind.NONE.equals(property.aggregation)
+                            ? context.pureModel.getEnumValue(M3Paths.AggregationKind, "None")
+                            : null
```
Comment (kevin-m-knight-gs) [inline legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/HelperModelBuilder.java:88]: If property.aggregation is a non-null and not one of the previous values, then we need to throw an exception.

**Review by kevin-m-knight-gs — CHANGES_REQUESTED** [review]
I think we need to be more rigorous about throwing exceptions when we run into an unexpected value. Also, we need some tests for invalid cases.

File: legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/to/HelperDomainGrammarComposer.java [hunk 3]
```diff
@@ -74,7 +75,18 @@ public static String renderMultiplicity(Multiplicity multiplicity)
...
+
+    private static String renderAggregation(AggregationKind aggregationKind)
+    {
+        return AggregationKind.NONE.equals(aggregationKind)
+                ? "(none) "
+                : AggregationKind.SHARED.equals(aggregationKind)
+                    ? "(shared) "
+                    : AggregationKind.COMPOSITE.equals(aggregationKind)
+                        ? "(composite) "
+                        : "";
```
[the remark this answers is not in the document]
Comment (akphi) [reply legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/to/HelperDomainGrammarComposer.java:89]: I think we avoid throwing error in grammar composer

File: legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/domain/DomainParseTreeWalker.java [hunk 4]
```diff
@@ -332,6 +333,15 @@ private Property visitSimpleProperty(DomainParserGrammar.PropertyContext ctx)
...
         property.sourceInformation = this.walkerSourceInformation.getSourceInformation(ctx);
         property.propertyTypeSourceInformation = this.walkerSourceInformation.getSourceInformation(ctx.propertyReturnType().type());
+        property.aggregation = ctx.aggregation() != null
+                ? ctx.aggregation().aggregationType().AGGREGATION_TYPE_COMPOSITE() != null
+                    ?  AggregationKind.COMPOSITE
+                    : ctx.aggregation().aggregationType().AGGREGATION_TYPE_SHARED() != null
+                        ? AggregationKind.SHARED
+                        : ctx.aggregation().aggregationType().AGGREGATION_TYPE_NONE() != null
+                            ? AggregationKind.NONE
+                            : null
```
[the remark this answers is not in the document]
Comment (akphi) [reply legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/domain/DomainParseTreeWalker.java:343]: I don't think we can do it here, it's part of ANTLR logic that has ensured that won't happen

[the remark this answers is not in the document]
Comment (akphi) [reply legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/domain/DomainParseTreeWalker.java:343]: Though I can definitely make this logic look clearer by not using ternary op here

[the remark this answers is not in the document]
Comment (akphi) [reply legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/domain/DomainParseTreeWalker.java]: Updated per suggestions, seem like I'm delusional to think we consistently don't throw error in composer 😢

File: legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/domain/DomainParseTreeWalker.java [hunk 5]
```diff
@@ -332,6 +333,19 @@ private Property visitSimpleProperty(DomainParserGrammar.PropertyContext ctx)
...
+            if (ctx.aggregation().aggregationType().AGGREGATION_TYPE_COMPOSITE() != null)
+            {
+                property.aggregation = AggregationKind.COMPOSITE;
+            } else if (ctx.aggregation().aggregationType().AGGREGATION_TYPE_SHARED() != null)
+            {
+                property.aggregation = AggregationKind.SHARED;
+            } else if (ctx.aggregation().aggregationType().AGGREGATION_TYPE_NONE() != null)
+            {
+                property.aggregation = AggregationKind.NONE;
+            }
```
Comment (kevin-m-knight-gs) [inline legend-engine-language-pure-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/domain/DomainParseTreeWalker.java]: I still think you need to throw an exception here if you get a non-null aggregation kind you do not understand. For example, what if someone adds a new aggregation kind in the parser grammar without updating this code? Then any usages of the new aggregation kind will be parsed incorrectly. It's much better to throw an exception to make sure you're not letting any cases fall through the cracks.
```suggestion
            if (ctx.aggregation().aggregationType().AGGREGATION_TYPE_COMPOSITE() != null)
            {
                property.aggregation = AggregationKind.COMPOSITE;
            }
            else if (ctx.aggregation().aggregationType().AGGREGATION_TYPE_SHARED() != null)
            {
                property.aggregation = AggregationKind.SHARED;
            }
            else if (ctx.aggregation().aggregationType().AGGREGATION_TYPE_NONE() != null)
            {
                property.aggregation = AggregationKind.NONE;
            }
            else
            {
                throw new EngineException("Unknown aggregation kind", EngineErrorType.PARSER);
            }
```

**Review by kevin-m-knight-gs — CHANGES_REQUESTED** [review]

**Review by kevin-m-knight-gs — APPROVED** [review]

File: legend-engine-language-pure-grammar/src/test/java/org/finos/legend/engine/language/pure/grammar/test/roundtrip/TestDomainGrammarRoundtrip.java (hunk 1 above)
Comment (pierredebelen) [reply legend-engine-language-pure-grammar/src/test/java/org/finos/legend/engine/language/pure/grammar/test/roundtrip/TestDomainGrammarRoundtrip.java:22]: Yep it's part of the UML metamodel

**Pull request merged**